### PR TITLE
Add SilverStripeComponent base class

### PR DESCRIPTION
**Pull request by flashbackzoo**
@chillu this is the event based interfacing we talked about today. It's still a bit coupled in terms of both sides (React component and legacy script) need to know about the namespace and event names...

Thoughts?

### CODE REVIEW

File: docs/en/silverstripe-component.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	componentDidMount() {
+		super.componentDidMount();
+	}
+	
+	componentWillUnmount() {
+		super.componentWillUnmount();
+	}
+	
+	componentDidUpdate() {
+		this._emitCmsEvent('updateCmsHeading', { heading: this.state.headingName });
```
Comment (chillu) [inline docs/en/silverstripe-component.md]: `my-component.title-changed`

File: docs/en/silverstripe-component.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+export default MyComponent;
+```
+
+__legacy.js__
+```javascript
+(function ($) {
+	$.entwine('ss', function ($) {
+		$('.cms-button').entwine({
+			onmatch: function () {
+				$(document).on('my-component.updateCmsHeading', function (e, data) {
```
Comment (chillu) [inline docs/en/silverstripe-component.md]: `$('.cms-breadcrumbs').text(data.title)`

File: docs/en/silverstripe-component.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+import $ from 'jquery';
+import React from 'react';
+import MyComponent from './my-component';
+
+$('.my-component-wrapper').entwine({
+	onadd: function () {
+		var props = {
+			cmsEventsNamespace: 'my-component',
+			cmsEvents: {
+				highlightComponent: function () {
```
Comment (chillu) [inline docs/en/silverstripe-component.md]: Include namespace here, remove `cmsEventsNamespace`
Use `cms.tabchanged` as a better example.

**Comment by flashbackzoo** [comment]
@chillu updated based on your feedback :smile:

**Comment by flashbackzoo** [comment]
@chillu I've updated the docs with the Entwine override stuff and squashed the commits. Ready for review and (hopefully) merge.

**Pull request merged**